Streamline PSA key type and curve encodings

gilles-peskine-arm opened the pull request:
Change `psa_key_type_t` to a more compact 16-bit type.

Change the encoding of elliptic curves (and FFDH groups) drastically: now there is an 8-bit curve family in the lower 8 bits of the key type. A curve family can include several curves as long as they all have different octet lengths. The precise curve is determined by the curve family and the key size, which is fully known when generating or deriving a key and known up to byte granularity when importing a key. Rationale: it makes the encoding more compact, and makes it easier for applications to work with multiple curves without having to embed a conversion between the curve/type encoding and the size.

Distinct key types have a Hamming distance of at least 2. This helps implementations that protect against a single bit flip. This is done by making the least significant bit a parity bit.

Elliptic curves no longer use the encoding from the TLS registry, which was restrictive because it has little room reserved for curves that aren't in the registry, but the world outside TLS uses quite a few curves that aren't used in TLS.

Private discussion about the design: https://github.com/ARMmbed/psa-crypto/issues/122#issuecomment-561096308

This breaks the build of the attestation service in Mbed OS due to the change in how elliptic curves are encoded.

Needs preceding PRs:

* https://github.com/ARMmbed/mbedtls/pull/2964 — change the code of ssl_cli so that it works with the new curve encoding.
* [upcoming PR to https://github.com/ARMmbed/mbed-os] — change the code of the attestation service so that it works with the new curve encoding.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -599,13 +601,9 @@
-        (type) == PSA_KEY_TYPE_ARC4 ? 1 :            \
-        (type) == PSA_KEY_TYPE_CHACHA20 ? 1 :            \
-        0)
+    (((type) & PSA_KEY_TYPE_CATEGORY_MASK) == PSA_KEY_TYPE_CATEGORY_SYMMETRIC ? \
```
Technically redundant test as the specification of the macro says "unspecified if not a supported cipher key type".

gilles-peskine-arm replied:
That's the Mbed Crypto implementation which currently tends to favor robustness over minimal code size.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -538,6 +611,14 @@
+ * 2048, 3072, 4096, 6144, 8192. A given implementation may support
+ * all of these sizes or only a subet.
+ */
+#define PSA_DH_GROUP_RFC7919            ((psa_dh_group_t) 0x02)
```
Presume this should bx `0x020000`? - to align with the definition of the ECC group families?

athoelke replied:
At least - when first introduced, until the commit that changes to 16-bit key types...

gilles-peskine-arm replied:
I amended the old commits.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -568,14 +581,22 @@
+ *
+ * This family includes groups with the following key sizes (in bits):
+ * 2048, 3072, 4096. A given implementation may support
+ * all of these sizes or only a subet.
```
"subet" -> "subset"

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -568,14 +581,22 @@
  *
  * This family includes groups with the following key sizes (in bits):
  * 2048, 3072, 4096, 6144, 8192. A given implementation may support
  * all of these sizes or only a subet.
```
"subet" -> "subset"

athoelke commented on include/psa/crypto_sizes.h:
```diff
@@ -198,38 +198,7 @@
-     (curve) == PSA_ECC_CURVE_CURVE25519       ? 255 : \
-     (curve) == PSA_ECC_CURVE_CURVE448         ? 448 : \
-     0)
+#define PSA_ECC_CURVE_BITS(curve) ((curve) & 0xffff)
```
This isn't going to work anymore after the removal of the curve size from the key type.

athoelke commented:
There are a few more places which refer to the ECC curves using wildcard references (PSA_ECC_CURVE_XXX):
- `PSA_ALG_ECDH`
- `psa_export_key()`

Patater commented on tests/suites/test_suite_psa_crypto_metadata.function:
```diff
@@ -56,6 +56,18 @@
+    value ^= value >> 4;
+    return( 0x9669 & 1 << ( value & 0xf ) );
+}
+#define TEST_PARITY( value )                    \
```
Suggestion: `TEST_ODD_PARITY()` to make it clear what about parity you are checking for.

gilles-peskine-arm replied:
It's actually even parity (so that 0 has the correct parity). And I think the macro name should not indicate which parity it is. The job of the macro is to test for the _expected_ parity, not for even parity.

Patater replied:
I'll explain where I'm coming from. If a hardware datasheet says "odd parity bit", I would expect a value of 1 there to mean odd parity and 0 to mean even parity. If a datasheet says "even parity bit", then I expect a value of 1 there to mean even parity, and 0 to mean odd parity. If the datasheet says "parity bit", I am unclear what the hardware will place there.

In this case, I read "value" as the value you wanted to check the parity of, not the expected parity. Consider renaming "value" to "expected_parity" if the job of the macro is to test for arbitrary expected parity. Consider renaming to `TEST_EVEN_PARITY` if value is the value you are checking the parity of. The comment you've added to `test_parity` is good, as it says what sort of parity you are checking for, but the names could still be more clear.

gilles-peskine-arm replied:
The name of the macro deliberately doesn't say which way the parity should go because its job is to test for whatever parity the API design intends. The fact that it's even parity is an implementation detail.

I don't understand how you come to name it odd parity. A value is correct iff the number of bits that is set is even.

I don't understand why you want to rename the argument. The argument is a value whose parity needs to be checked. There is no argument indicating what parity is expected: that's built into the code, since all values that are checked for parity must obey the same parity requirement.

gilles-peskine-arm commented on include/mbedtls/psa_util.h:
```diff
@@ -160,81 +160,96 @@ static inline psa_algorithm_t mbedtls_psa_translate_md( mbedtls_md_type_t md_alg
-            *oid = MBEDTLS_OID_EC_GRP_BP512R1;
-            *oid_len = MBEDTLS_OID_SIZE( MBEDTLS_OID_EC_GRP_BP512R1 );
-            return( 0 );
+                case 521:
```
Typo. Fixed by amending the commit.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -375,71 +375,118 @@ static inline int psa_key_slot_is_external( const psa_key_slot_t *slot )
+            return( PSA_ECC_CURVE_BRAINPOOL_P_R1 );
         case MBEDTLS_ECP_DP_CURVE25519:
-            return( PSA_ECC_CURVE_CURVE25519 );
+            *bits = 519;
```
My search-and-replace wasn't quite right. Fixed by amending the commit.

gilles-peskine-arm commented on include/psa/crypto_types.h:
```diff
@@ -63,85 +63,29 @@ typedef int32_t psa_status_t;
- * 0xCACA
- * 0xDADA
- * 0xEAEA
- * 0xFAFA
```
GREASE is a TLS feature. It's irrelevant here. Using values reserved for GREASE was a hack to get more room for values not defined by the TLS registry. Since we no longer leverage the TLS registry, GREASE is not relevant anymore.

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -271,5 +271,18 @@ ECDSA restartable sign/verify: ECKEY, max_ops=250
+
+PSA wrapped sign: BP256R1
+depends_on:MBEDTLS_ECP_DP_BP256R1_ENABLED
+pk_psa_sign:MBEDTLS_ECP_DP_BP256R1:PSA_ECC_CURVE_BRAINPOOL_P_R1:256
```
I added a test for all supported curves in Mbed Crypto [except secp224k1](https://github.com/ARMmbed/mbed-crypto/issues/336). There's no test for Curve25519 because it can't be used for ECDSA and that's the only pk-over-psa interface that's implemented.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -5217,6 +5279,10 @@ static psa_status_t psa_key_agreement_ecdh( const uint8_t *peer_key,
                                   mbedtls_ctr_drbg_random,
                                   &global_data.ctr_drbg ) );
+    if( status != PSA_SUCCESS )
+        goto exit;
```
Good point. `mbedtls_ecdh_calc_secret` is implemented safely (it only writes its output as the very last, non-fallible step), but robustness here is good. More importantly, if the size check fails, it's possible that the leaked output would be sensitive (I haven't analyzed to check).

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -271,5 +271,18 @@ ECDSA restartable sign/verify: ECKEY, max_ops=250
-pk_psa_sign:
+PSA wrapped sign: SECP256R1
+depends_on:MBEDTLS_ECP_DP_SECP256R1_ENABLED
+pk_psa_sign:MBEDTLS_ECP_DP_SECP256R1:PSA_ECC_CURVE_SECP_R1:256
```
No, pk-over-psa is not implemented for RSA.

gilles-peskine-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -1228,15 +1233,23 @@ exit:
-/* BEGIN_CASE depends_on:MBEDTLS_SHA256_C:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED */
-void pk_psa_sign(  )
+/* BEGIN_CASE depends_on:MBEDTLS_SHA256_C:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_ECDSA_C */
+void pk_psa_sign( int grpid_arg,
```
This is a preexisting function. Renaming it would be out of scope.

Patater replied:
Could easily be added in a new commit within the same PR. Next opportunity for rename is when we add a new function for testing RSA.

gilles-peskine-arm replied:
Or we might modify this function to support RSA, in which case renaming it now would have been the wrong thing. This PR is big enough as it is; I'm not renaming this function.

Patater replied:
Making the function do even more and be even larger would be dumb. Test functions should be small and focused on what they are testing.

gilles-peskine-arm replied:
The _role_ of the function is to test signing through pk-over-psa. The current implementation is a monolithic sequence of actions to generate a key, encapsulate it, perform the signature and verify the result. Good, textbook design would be to make each step into an auxiliary function that receives the key type as an argument (if necessary) and does its job given the key type. Having separate functions with the same structure for different key types would violate DRY. Now, these are unit tests, and some amount of repetition is excusable. Still, I would prefer to have more auxiliary functions and less repetition in our test suites, not more.

AndrzejKurek commented on include/psa/crypto_compat.h:
```diff
@@ -50,6 +50,8 @@ extern "C" {
 typedef MBEDTLS_PSA_DEPRECATED size_t mbedtls_deprecated_size_t;
```
Minor: typo in commit 27f8944 - `Remove the values of curve encodings that is based` - encodings/values are plural, so I'd assume that it should be that **are** based.
Also, `andinclude` -> `and include`

gilles-peskine-arm replied:
I fixed the commit message.

AndrzejKurek commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -2314,79 +2314,79 @@ derive_key:PSA_ALG_HKDF(PSA_ALG_SHA_512):"0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
 PSA raw key agreement: ECDH SECP384R1 (RFC 5903)
 depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_ECP_C:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_ECDH_C
-raw_key_agreement:PSA_ALG_ECDH:PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP384R1):"099f3c7034d4a2c699884d73a375a67f7624ef7c6b3c0f160647b67414dce655e35b538041e649ee3faef896783ab194":"04e558dbef53eecde3d3fccfc1aea08a89a987475d12fd950d83cfa41732bc509d0d1ac43a0336def96fda41d0774a3571dcfbec7aacf3196472169e838430367f66eebe3c6e70c416dd5f0c68759dd1fff83fa40142209dff5eaad96db9e6386c":"11187331c279962d93d6 ...
+raw_key_agreement:PSA_ALG_ECDH:PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_CURVE_SECP_R1):"099f3c7034d4a2c699884d73a375a67f7624ef7c6b3c0f160647b67414dce655e35b538041e649ee3faef896783ab194":"04e558dbef53eecde3d3fccfc1aea08a89a987475d12fd950d83cfa41732bc509d0d1ac43a0336def96fda41d0774a3571dcfbec7aacf3196472169e838430367f66eebe3c6e70c416dd5f0c68759dd1fff83fa40142209dff5eaad96db9e6386c":"11187331c279962d93d604 ...
```
I assume that key bits can be inferred from key data?

gilles-peskine-arm replied:
Yes, in most cases. The exception would be for EC or DH keys if there were two curves in the same family with sizes that are within the same range [8*k, 8*k+7], but there are no such families and I don't expect there to ever be any such family.

Patater commented on include/psa/crypto_values.h:
```diff
@@ -458,70 +458,82 @@
+ *   _Ed448-Goldilocks, a new elliptic curve_, NIST ECC Workshop, 2015.
+ *   The algorithm #PSA_ALG_ECDH performs X448 when used with this curve.
+ */
+#define PSA_ECC_CURVE_XED               ((psa_ecc_curve_t) 0x41)
```
Consider using `PSA_ECC_FAMILY_MONTGOMERY` instead, to be more up to date with the latest PSA spec.

gilles-peskine-arm replied:
Since I was rewriting history anyway, I made it `MONTGOMERY` from the get go.

gilles-peskine-arm commented:
I rebased yet again.

* Rebased on top of `development` so that there are no merge conflicts.
* Fixed typos in older commits.
* Rewrote history so that it's `PSA_ECC_CURVE_MONTGOMERY` from the start and not `_XED`.

gilles-peskine-arm commented:
The Crypto CI passed except for Mbed OS which is already broken before this PR and will require a further update after this PR.

The TLS CI is not expected to pass here because an update of the crypto submodule is required. https://github.com/ARMmbed/mbedtls/pull/2964 updates the crypto submodule.

gilles-peskine-arm commented:
Another day, another merge conflict. I'll rebase.

gilles-peskine-arm commented:
The Crypto CI passed except for Mbed OS which is already broken before this PR and will require a further update after this PR.

The TLS CI is not expected to pass here because an update of the crypto submodule is required. https://github.com/ARMmbed/mbedtls/pull/2964 updates the crypto submodule.

gilles-peskine-arm commented:
With two approvals and with acceptable CI (crypto and TLS pass except for known unrelated failures), this is ready for merge. Since it will disrupt TLS testing until the corresponding TLS PR https://github.com/ARMmbed/mbedtls/pull/2964 is merged, we should merge this when we're ready to move on with https://github.com/ARMmbed/mbedtls/pull/2964.
